app/ui-message/client/actionButtons/lib/applyButtonFilters.ts

```diff
@@ -0,0 +1,37 @@
+/* Style disabled as having some arrow functions in one-line hurts readability */
+/* eslint-disable arrow-body-style */
+
+import { Meteor } from 'meteor/meteor';
+import { IUIActionButton, TemporaryRoomTypeFilter } from '@rocket.chat/apps-engine/definition/ui';
+
+import { hasAtLeastOnePermission, hasPermission, hasRole } from '../../../../authorization/client';
+import { IRoom, RoomType } from '../../../../../definition/IRoom';
+
+export const applyAuthFilter = (button: IUIActionButton, room?: IRoom): boolean => {
+	const { hasAllPermissions, hasOnePermission, hasAllRoles, hasOneRole } = button.when || {};
+
+	const hasAllPermissionsResult = hasAllPermissions ? hasPermission(hasAllPermissions) : true;
+	const hasOnePermissionResult = hasOnePermission ? hasAtLeastOnePermission(hasOnePermission) : true;
+	const hasAllRolesResult = hasAllRoles ? hasAllRoles.every((role) => hasRole(Meteor.userId(), role, room?._id)) : true;
+	const hasOneRoleResult = hasOneRole ? hasRole(Meteor.userId(), hasOneRole, room?._id) : true;
+
+	return hasAllPermissionsResult && hasOnePermissionResult && hasAllRolesResult && hasOneRoleResult;
+};
+
+const mapFilterToRoomType: {[k in TemporaryRoomTypeFilter]: RoomType} = {
+	[TemporaryRoomTypeFilter.CHANNEL]: 'c',
+	[TemporaryRoomTypeFilter.GROUP]: 'p',
+	[TemporaryRoomTypeFilter.DIRECT]: 'd',
+	[TemporaryRoomTypeFilter.DIRECT_MULTIPLE]: 'd',
+	[TemporaryRoomTypeFilter.TEAM]: 'p',
+	[TemporaryRoomTypeFilter.LIVE_CHAT]: 'l',
+};
```

Comment:
The enum `TemporaryRoomTypeFilter` was based off of [this property](https://github.com/RocketChat/Rocket.Chat/blob/feat/add-ui-buttons-to-contexts/client/views/room/lib/Toolbox/index.tsx#L31), but not all values have equivalents in the `RoomType` [type here](https://github.com/RocketChat/Rocket.Chat/blob/feat/add-ui-buttons-to-contexts/definition/IRoom.ts#L5)

Is `RoomType` incomplete? Or the values in `ToolboxActionConfig['groups']` are not supposed to map to room types?

Reply:
Yes, because it's result of a runtime-based declarative API (`roomTypes.add(config: RoomTypeConfig)`). It might change in future. Ask @RocketChat/chat-engine for more info on it.

Reply:
Add new types for filtering:
```
RoomTypeFilter = [ PRIVATE_CHANNEL, PUBLIC_CHANNEL, DIRECT, DIRECT_MULTIPLE, PUBLIC_TEAM, PRIVATE_TEAM, PRIVATE_DISCUSSION, PUBLIC_DISCUSSION, LIVE_CHAT ]
```